### PR TITLE
[doc] Refocus cryptolib documentation text.

### PR DESCRIPTION
**Pull request by jadephilipoom**
Make the tone a little more conversational and change the focus from background information to practical usage.

In #19297 I moved the working cryptolib API document onto a public documentation page. I didn't change any text at the time, to minimize the size of the change. But I think it makes sense to shift the focus of the text a little bit now, since the purpose of a public documentation page (communicate how to use the crypto library) is different from an internal document (to provide background, explain why certain things are included, and agree on what to implement). So I went through the text and made some edits:
- adding additional explanation where I think it is helpful for usage (e.g. explaining how the key structs work, specific security considerations for ECC/RSA)
- shortening prose to make the page easier to browse (e.g. removing exposition that could be found through reference links or in other sections)
- adding cross-links to other documentation pages or sections where appropriate

I just made the edits that subjectively made sense to me for the new format; feedback is of course very welcome :slightly_smiling_face: 

CC @hcallahan-lowrisc @moidx for visibility

### CODE REVIEW

File: doc/security/cryptolib/cryptolib_api.md [hunk 1]
```diff
@@ -1,149 +1,123 @@
...
+| [**Message authentication**](#message-authentication) | HMAC-SHA256<br>KMAC{128,256} |
+| [**RSA**](#rsa) | RSA-{2048,3072,4096} |
+| [**Elliptic curve cryptography**](#elliptic-curve-cryptography) | ECDSA-{P256,P384}<br>ECDH-{P256,P384}<br>Ed25519<br>X25519 |
+| [**Deterministic random bit generation**](#deterministic-random-bit-generation) | AES-CTR-DRBG |
+| [**Key derivation**](#key-derivation) | HMAC-KDF-CTR<br>KMAC-KDF-CTR |
+
+
+## Asynchronous operations
+
+For some functions, OpenTitan's cryptolib support asynchronous calls.
```
Comment (jwnrt) [inline doc/security/cryptolib/cryptolib_api.md:27]:
```suggestion
For some functions, OpenTitan's cryptolib supports asynchronous calls.
```

**Review by jwnrt — CHANGES_REQUESTED** [review]
Thanks @jadephilipoom, this is really high quality documentation.

I have only checked the syntax/formatting/prose and trust that the content is correct (and in theory hasn't changed).

Given how much content there is in this page, it may or may not be easier to browse broken into more pages, but fine to leave it as is here.

**Comment by jadephilipoom** [comment]
Thanks for catching all my typos @jwnrt ! They should all be fixed now.

**Review by jwnrt — APPROVED** [review]

**Review by moidx — APPROVED** [review]

**Comment by hcallahan-lowrisc** [comment]
This looks great, I think the new tone is much more approachable for users. Thanks!

I have some very general feedback which is not urgent at all to address.
- To me, a specification is more terse and is more about requirements than implementation details. Maybe the title should be "OpenTitan Cryptography Library User Guide"? or API Guide?
- I think the introduction could be fleshed out a little. As a reader I feel a little disoriented dropping straight into "Asynchronous operations" after the Algo table. Even with the rhs ToC when hosted on the site, it's a bit overwhelming. Maybe after the first paragraph, you could have some bullet points / links along the lines of :
> This page describes the following functionality:
> - Asynchronous Calling Conventions
> - API Data Structures / Types
> - DRBG
> - Key Derivation and Import/Export
> - Supported Algorithms & Primitive Operations
- It would also be nice to have a small table somewhere near the top (maybe after the algos table) with links to all the Crypto HWIP blocks in one place. It would help the reader with the thought e.g. "how/where is X implemented? Oh there's links to everything at the top of the page, I'll start there".
- The section on security strength could probably go on it's own page, maybe with a duplication of the table of supported algorithms

**Comment by jadephilipoom** [comment]
Thanks for the feedback! I like these ideas and will make a follow-up PR to implement them :slightly_smiling_face:

File: doc/security/cryptolib/cryptolib_api.md [hunk 2]
```diff
@@ -1,149 +1,123 @@
...
+    - Takes caller-allocated output buffers. Blocks until OTBN is done
+      processing if needed, then checks whether it had errors. If not, does any
+      necessary postprocessing and writes results to the buffers.
+
+The caller should call the `start` function, wait for the interrupt, and then call `finalize`.
+
+A few noteworthy aspects of this setup:
+- While an asynchronous operation is running, OTBN will be unavailable and attempts to use it will return errors.
+- Only one asynchronous operation may be in progress at any given time.
+- The caller is responsible for properly managing asynchronous calls, including ensuring that the entity receiving the `finalize` results is the same as the one who called `start`.
```
Comment (cdgori) [inline doc/security/cryptolib/cryptolib_api.md:51]: Thanks for these changes @jadephilipoom - everything is great in terms of the refocusing of the documentation and I think it makes this much more user-friendly.

One small comment about this section here - do we want to expand on this and say what happens (or how we might mitigate against) an incorrect `finalize` call here?  (Either maliciously placed or inadvertent/bug-induced?)

Comment (jadephilipoom) [reply doc/security/cryptolib/cryptolib_api.md:51]: Some of that depends on the specifics of the OTBN program and the async algorithm that was called. There are some library-level mitigations we could do -- some ideas on https://github.com/lowRISC/opentitan/issues/19173 but none of these are yet implemented. Right now we don't really mitigate against this case and, if an attacker or another application calls `finalize()` before the original caller and no one runs a new OTBN operation after that, the original caller will silently get random data when they call `finalize()`.

**Pull request merged**